Migrate from vue-query 1.x to @tanstack/vue-query 4.x

agualis opened the pull request:
# Description

`vue-query` v2 was released sometime ago.
Furthermore,  `vue-query` was migrated to be a part of https://github.com/TanStack/query. 

This PR migrates to the latest version, which involves this [breaking changes](https://vue-query-next.vercel.app/#/guides/migrating-to-vue-query-2): 

- `VueQueryDevTools` was removed. You can simply use official vue devtools instead.
- Callback methods like `refetch` or `fetchNextPage` are not refs anymore so no need to call them with `.value()`.
- Typescript types were improved. No more references to `react-query`.
- `idle` state was removed. Our use cases (no offline mode) just need `isLoading`, which simplifies logic. 
- Setup is done through a plugin which will simplify incoming global providers refactor.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## How should this be tested?

Please provide instructions so we can test. Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

## Visual context

Please provide any relevant visual context for UI changes or additions. This could be static screenshots or a loom screencast.

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have requested at least 1 review (If the PR is significant enough, use best judgement here)
- [ ] I have commented my code where relevant, particularly in hard-to-understand areas
- [ ] If package-lock.json has changes, it was intentional.
- [ ] The base of this PR is `master` if hotfix, `develop` if not

agualis commented on src/composables/queries/useGraphQuery.ts:
```diff
@@ -63,5 +62,5 @@ export default function useGraphQuery<T>(
   };
 
-  return useQuery(queryKey, queryFn, options);
+  return useQuery(queryKey, queryFn, options as any);
```
I did not find a way to make TS compiling the options type 🤷‍♂️
Maybe you can give it a try 🧙

sssmi replied:
How's this:
```suggestion
  return useQuery<T>(queryKey, queryFn, options);
```

agualis replied:
I forgot that useQuery was expecting a generic.  Thanks!

agualis commented on src/providers/local/staking/userUserStakingData.ts:
```diff
@@ -59,63 +57,15 @@ export type PoolGaugesResponse = {
   pools: PoolGaugesInfo[];
 };
 
-export type UserStakingDataResponse = {
```
[Using TS type inference with ReturnType](https://github.com/balancer-labs/frontend-v2/pull/2608/files#diff-431f5029dfdaefe2a611e7169e9fb4b62bbe00314e90efb9c8dfde20b5acffd0) avoids manually maintaining this complex type. The only drawback is that I had to move the related comments close to the related declarations. In any case, most of those comments should not be needed in case of proper naming.

sssmi commented on src/providers/local/staking/userUserStakingData.ts:
```diff
@@ -307,7 +264,7 @@ export default function useUserStakingData(
   );
 
-  const { data: poolBoosts, isLoading: isLoadingBoosts } = useQuery<
+  const { data: poolBoosts, isInitialLoading: isLoadingBoosts } = useQuery<
```
Have to be careful with queries that are not always enabled like this one here. Their `isLoading` state stays `false` until they are enabled. This caused the `StakePreview.vue` to show perpetual loading block.

<img width="1070" alt="Screenshot 2023-01-05 at 14 55 35" src="https://user-images.githubusercontent.com/28311328/210785321-f09977b4-738b-4c73-8728-25e90de89475.png">

Checking the `isInitialLoading` instead works here. https://tanstack.com/query/v4/docs/react/guides/disabling-queries

agualis replied:
I didn't catch that braking change in the docs. I think that we will have more cases like this one.

simeonkerkola reviewed (approved):
This is a good upgrade. Seems to be working fine 👍

agualis commented:
@garethfuller @simeonkerkola I fixed conflicts and updated the queries in the staking refactor. I think that we should try to merge this before we add/update more queries but we should: 

1. Double check that we don't need more adjustments like the ones Simeon did in [this commit](https://github.com/balancer-labs/frontend-v2/pull/2608/commits/eb739bfa13c509eecf792ef14b45c7499c78ea94) 
2. Fix `tokens.provider` tests as they were affected by some refactor in the related `dynamicDataLoaded.value` logic.

Could you help me with this 2 points as you are more familiar with this?

garethfuller commented:
Closing in favour of #2835
